src/main/java/com/spotify/autoscaler/Autoscaler.java

```diff
@@ -85,33 +88,40 @@ public AutoscaleJob makeAutoscaleJob(
 
   @Override
   public void run() {
-    /*
-     * Without this horrible bit of horribleness,
-     * any uncaught Exception would kill the whole autoscaler.
-     */
+    Future<?> runResult = executorService.submit(this::check);
     try {
-      runUnsafe();
-    } catch (final Exception t) {
-      LOGGER.error("Unexpected Exception!", t);
+      runResult.get(THREADS_TIMEOUT, TimeUnit.MILLISECONDS);
+    } catch (TimeoutException e) {
+      LOGGER.warn("Autoscaler is taking too long to run, stopping this check.");
+      runResult.cancel(true);
+    } catch (final Throwable t) {
+      LOGGER.error("Exception happened. Shutting down the whole application.", t);
+      executorService.shutdownNow();
+      System.exit(1);
     }
   }
 
-  private void runUnsafe() {
+  private void check() {
+    LOGGER.info("Starting Autoscaler check.");
     autoscalerMetrics.markHeartBeat();
+    List<BigtableCluster> candidateClusters = database.getCandidateClusters();
+    LOGGER.info("Got {} candidate clusters from the database.", candidateClusters.size());
+
     final CompletableFuture[] futures =
-        database
-            .getCandidateClusters()
+        candidateClusters
             .stream()
             // Order here is important - don't call updateLastChecked if a cluster is filtered.
             // That could lead to cluster starvation
             .filter(filter::match)
             .filter(database::updateLastChecked)
             .map(
                 cluster ->
-                    CompletableFuture.runAsync(() -> runForCluster(cluster), executorService))
+                    CompletableFuture.runAsync(() -> runForCluster(cluster), executorService)
+                        .orTimeout(THREADS_TIMEOUT, TimeUnit.MILLISECONDS))
```

Comment:
do we need to have timeouts for each task? i think one timeout for the whole process would suffice, but if we want to have individual timeouts, probably having the whole process ~30 seconds and individual tasks ~5 seconds would be good

Reply:
Probably not, was an additional layer of safety.